File: nodns-resolver/src/dns.ts
import type { DnsRecordType, DohAnswer, DohResponse } from './types.js';
import { DEFAULT_DOH_ENDPOINT, DNS_TYPE_MAP } from './types.js';

const DOH_TIMEOUT_MS = 10_000;

export async function queryDoh(
  fqdn: string,
  type: string,
  dohEndpoint: string = DEFAULT_DOH_ENDPOINT,
): Promise<DohResponse> {
  const url = `${dohEndpoint}?name=${encodeURIComponent(fqdn)}&type=${type}`;

  const controller = new AbortController();
  const timer = setTimeout(() => controller.abort(), DOH_TIMEOUT_MS);

  let response: Response;
  try {
    response = await fetch(url, {
      headers: { Accept: 'application/dns-json' },
      signal: controller.signal,
    });
  } catch (err: unknown) {
    if (err instanceof DOMException && err.name === 'AbortError') {
      throw new Error('DNS query timed out. Please try again.');
    }
    throw new Error('DNS query failed. Please check your network.');
  } finally {
    clearTimeout(timer);
  }

  if (!response.ok) {
    throw new Error(`DNS query failed (HTTP ${response.status}).`);
  }

  return (await response.json()) as DohResponse;
}

function stripTxtQuotes(data: string): string {
  return data.replace(/^"/, '').replace(/"$/, '').replace(/"\s*"/g, '');
}

function dnsTypeNumberToString(num: number): string {
  return DNS_TYPE_MAP[num] ?? String(num);
}

export async function queryDnsRecords(
  name: string,
  type: DnsRecordType,
  dohEndpoint: string = DEFAULT_DOH_ENDPOINT,
): Promise<string[]> {
  const resp = await queryDoh(name, type, dohEndpoint);
  if (!resp.Answer) return [];

  return resp.Answer.filter(
    (a: DohAnswer) => dnsTypeNumberToString(a.type) === type,
  ).map((a: DohAnswer) => (type === 'TXT' ? stripTxtQuotes(a.data) : a.data));
}

export async function queryAllDnsRecordTypes(
  fqdn: string,
  types: readonly DnsRecordType[] = ['A', 'AAAA', 'TXT', 'CNAME', 'MX'],
  dohEndpoint: string = DEFAULT_DOH_ENDPOINT,
): Promise<{ name: string; type: string; ttl: number; data: string }[]> {
  const results: { name: string; type: string; ttl: number; data: string }[] = [];

  for (const type of types) {
    try {
      const resp = await queryDoh(fqdn, type, dohEndpoint);
      if (!resp.Answer) continue;
      for (const a of resp.Answer) {
        const typeStr = dnsTypeNumberToString(a.type);
        if (typeStr !== type) continue;
        results.push({
          name: a.name,
          type: typeStr,
          ttl: a.TTL,
          data: type === 'TXT' ? stripTxtQuotes(a.data) : a.data,
        });
      }
    } catch {
      // individual type failures are non-fatal
    }
  }

  return results;
}

File: nodns-resolver-rs/src/dns.rs
//! DNS-over-HTTPS (DoH) lookup functions.

use std::time::Duration;

use anyhow::Result;
use url::Url;

use crate::types::{dns_type_number_to_string, DnsAnswer, DohResponse, DEFAULT_DNS_TYPES};

const DOH_TIMEOUT_SECS: u64 = 10;

pub async fn query_doh(
    fqdn: &str,
    record_type: &str,
    doh_endpoint: &str,
    http: &reqwest::Client,
) -> Result<DohResponse> {
    let mut url = Url::parse(doh_endpoint)?;
    url.query_pairs_mut()
        .append_pair("name", fqdn)
        .append_pair("type", record_type);

    let resp = http
        .get(url)
        .header("Accept", "application/dns-json")
        .timeout(Duration::from_secs(DOH_TIMEOUT_SECS))
        .send()
        .await?;

    if !resp.status().is_success() {
        anyhow::bail!("DNS query failed (HTTP {}).", resp.status());
    }

    let doh: DohResponse = resp.json().await?;
    Ok(doh)
}

fn strip_txt_quotes(data: &str) -> String {
    data.trim_start_matches('"')
        .trim_end_matches('"')
        .replace("\" \"", "")
        .replace("\"", "")
}

pub async fn query_all_dns_record_types(
    fqdn: &str,
    types: Option<&[&str]>,
    doh_endpoint: &str,
    http: &reqwest::Client,
) -> Vec<DnsAnswer> {
    let types = types.unwrap_or(DEFAULT_DNS_TYPES);
    let mut results = Vec::new();

    for &rt in types {
        if let Ok(resp) = query_doh(fqdn, rt, doh_endpoint, http).await {
            if let Some(answers) = resp.answer {
                for a in answers {
                    let type_str = dns_type_number_to_string(a.type_num);
                    if type_str != rt {
                        continue;
                    }
                    let data = if rt == "TXT" {
                        strip_txt_quotes(&a.data)
                    } else {
                        a.data.clone()
                    };
                    results.push(DnsAnswer {
                        name: a.name,
                        record_type: type_str,
                        ttl: a.ttl,
                        data,
                    });
                }
            }
        }
    }

    results
}

pub async fn query_all_dns_record_types_default(
    fqdn: &str,
    doh_endpoint: &str,
    http: &reqwest::Client,
) -> Vec<DnsAnswer> {
    query_all_dns_record_types(fqdn, None, doh_endpoint, http).await
}
